shared/language-specs/languages.ts

```diff
@@ -23,128 +22,149 @@ const clojureSpec: LanguageSpec = {
     languageID: 'clojure',
     stylized: 'Clojure',
     fileExts: ['clj', 'cljs', 'cljx'],
-    identCharPattern: createIdentCharPattern('-!?+*<>='),
-    commentStyle: lispStyleComment,
+    identCharPattern: createIdentCharPattern('\\-!?+*<>='),
+    commentStyles: [lispStyleComment],
 }
 
 const csharpSpec: LanguageSpec = {
     languageID: 'csharp',
     stylized: 'C#',
     fileExts: ['cs', 'csx'],
-    commentStyle: cStyleComment,
+    commentStyles: [cStyleComment],
 }
 
 const dartSpec: LanguageSpec = {
     languageID: 'dart',
     stylized: 'Dart',
     fileExts: ['dart'],
-    commentStyle: { lineRegex: tripleSlashPattern },
+    commentStyles: [{ lineRegex: tripleSlashPattern }],
 }
 
 const elixirSpec: LanguageSpec = {
     languageID: 'elixir',
     stylized: 'Elixir',
     fileExts: ['ex', 'exs'],
     identCharPattern: rubyIdentCharPattern,
-    commentStyle: {
-        ...pythonStyleComment,
-        docPlacement: 'above the definition',
-    },
-    docstringIgnore: leadingAtSymbolPattern,
+    commentStyles: [
+        {
+            ...pythonStyleComment,
+            docPlacement: 'above the definition',
+            docstringIgnore: leadingAtSymbolPattern,
+        },
+    ],
 }
 
 const erlangSpec: LanguageSpec = {
     languageID: 'erlang',
     stylized: 'Erlang',
     fileExts: ['erl'],
-    commentStyle: {
-        // %% comment
-        lineRegex: /%%\s?/,
-    },
-    docstringIgnore: /-spec/,
+    commentStyles: [
+        {
+            // %% comment
+            lineRegex: /%%\s?/,
+            // -spec id(X) -> X.
+            docstringIgnore: /^\s*-spec/,
+        },
+    ],
 }
 
 const graphqlSpec: LanguageSpec = {
     languageID: 'graphql',
     stylized: 'GraphQL',
     fileExts: ['graphql'],
-    commentStyle: shellStyleComment,
+    commentStyles: [shellStyleComment],
 }
 
 const groovySpec: LanguageSpec = {
     languageID: 'groovy',
     stylized: 'Groovy',
     fileExts: ['groovy'],
-    commentStyle: cStyleComment,
+    commentStyles: [cStyleComment],
 }
 
+// {-# PRAGMA args #-}
+const haskellPragma = /^\s*{-#\s*[A-Z]+.*#-}$/
+
 const haskellSpec: LanguageSpec = {
     languageID: 'haskell',
     stylized: 'Haskell',
     fileExts: ['hs', 'hsc'],
     identCharPattern: createIdentCharPattern("'"),
-    commentStyle: {
-        // -- comment
-        // -- | doc comment
-        // {- block comment -}
-        // TODO - (support -- ^ doc comment)
-        lineRegex: /--\s?\|?\s?/,
-        block: { startRegex: /{-/, endRegex: /-}/ },
-    },
-    docstringIgnore: /INLINE|^#/,
+    commentStyles: [
+        {
+            // -- comment
+            // -- | doc comment
+            // {- block comment -}
+            lineRegex: /--\s?\|?\s?/,
+            block: { startRegex: /{-/, endRegex: /-}/ },
+            docstringIgnore: haskellPragma,
+        },
+        {
+            // -- ^ doc comment
+            lineRegex: /--\s?\^?\s?/,
+            docPlacement: 'above the definition',
```

Comment:
I think this is backward, isn't it? 

```haskell
f :: A -> B
-- ^Do a thing (after)
```

Or does "above" mean relative to the comment?

Reply:
This is why we have code reviews.